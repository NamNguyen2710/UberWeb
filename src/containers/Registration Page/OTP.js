import {useState} from 'react';

const Otp = (props) => {
  console.log(props)
  const [otp, setOtp] = useState(new Array(4).fill(''));
  const handleChange = (element, index) => {
    setOtp([...otp.map((val, idx) => idx === index ? element.value : val)]);
    if (element.nextSibling) {
      element.nextSibling.focus();
    }
  }
  return (
    <div className="signupForm">
      <div className="col">
        <p>Enter the 4-digit code sent to you at {props.phoneNumber}. </p>
        <p className="clickable-line" onClick={props.revertStep} >Did you enter the correct number?</p>
        <div className="otp-row">
          { otp.map((value, index) => 
            <input 
              className="otp-box"
              type="text"
              name="otp"
              autoComplete="off"
              maxLength="1"
              key={index}
              value={value}
              onChange={event => handleChange(event.target, index)}
            />
          )}
        </div>
        <p className="clickable-line">I didn’t receive code.</p>
        <button className="round-btn" onClick={props.handleSubmit}>→</button>
      </div>
    </div>
  );
}

export default Otp;
