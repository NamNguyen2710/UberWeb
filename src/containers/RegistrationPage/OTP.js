import {useState} from 'react';

const Otp = (props) => {
  const [otp, setOtp] = useState(new Array(4).fill(''));
  
  const handleChange = (element, index) => {
    setOtp([...otp.map((val, idx) => idx === index ? element.value : val)]);
    
  }
  const inputFocus = (event) => {
    if (event.key === "Delete" || event.key === "Backspace") {
      if (event.target.previousSibling) {
        event.target.previousSibling.focus()
      }
    } else {
      if (event.target.nextSibling) {
        event.target.nextSibling.focus();
      }
    }
  };
  const handleSubmit = () => { props.handleSubmit(otp) };

  return (
    <div className="signupForm">
      <div className="col">
        <p>Enter the 4-digit code sent to you at {props.phoneNumber}. </p>
        <p className="clickable-line" onClick={props.revertStep} >Did you enter the correct number?</p>
        <div className="otp-row">
          { otp.map((value, index) => 
            <input 
              className="otp-box"
              type="text"
              name="otp"
              autoComplete="off"
              maxLength="1"
              key={index}
              value={value}
              onChange={event => handleChange(event.target, index)}
              onKeyUp={event => inputFocus(event, index)}
            />
          )}
        </div>
        <p className="error-msg">{props.error}</p>
        <p className="clickable-line">I didn’t receive code.</p>
        <button className="round-btn" onClick={handleSubmit}>→</button>
      </div>
    </div>
  );
}

export default Otp;
