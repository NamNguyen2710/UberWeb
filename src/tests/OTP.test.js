import React, { useState } from "react";
import toJson from "enzyme-to-json";
import { shallow, mount } from "enzyme";
import { waitFor } from "@testing-library/react";
import OTP from "../containers/RegistrationPage/OTP";

describe("sign up otp form", () => {
  it("render correctly", () => {
    const handleSubmit = jest.fn();
    const revertStep = jest.fn();
    const phoneNumber = "123 456 7890";
    const error = "Invalid";

    const wrapper = shallow(
      <OTP
        handleSubmit={handleSubmit}
        revertStep={revertStep}
        phoneNumber={phoneNumber}
        error={error}
      />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("submit incorrect otp", async () => {
    const handleSubmit = jest.fn();
    const revertStep = jest.fn();
    const phoneNumber = "123 456 7890";

    const wrapper = mount(
      <OTP
        handleSubmit={handleSubmit}
        revertStep={revertStep}
        phoneNumber={phoneNumber}
      />
    );

    wrapper.find("button").first().simulate("click");
    await waitFor(() => {
      expect(handleSubmit).not.toHaveBeenCalled();
      expect(wrapper.find(".error-msg").text()).toEqual("Incorrect OTP");
    });
  });

  it.only("submit correct otp", async () => {
    const handleSubmit = jest.fn();
    const revertStep = jest.fn();
    const phoneNumber = "123 456 7890";
    const event = { target: { nextSibling: { focus: jest.fn() }, value: "5" } };

    const wrapper = mount(
      <OTP
        handleSubmit={handleSubmit}
        revertStep={revertStep}
        phoneNumber={phoneNumber}
      />
    );

    wrapper.find(".clickable-line").first().simulate("click");
    await waitFor(() => {
      expect(revertStep).toHaveBeenCalled();
    });

    for (let i = 0; i++; i <= 3) {
      wrapper.find('input[name="otp"]').at(i).props().onChange(event);
      wrapper.update();
      await waitFor(() => {
        expect(wrapper.find('input[name="otp"]').at(i).props().value).toBe("5");
        expect(event.target.nextSibling.focus).toHaveBeenCalledTimes(i + 1);
      });
    }

    wrapper.find("button").simulate("click");
    await waitFor(() => {
      expect(handleSubmit).toHaveBeenCalled();
    });
  });
});
