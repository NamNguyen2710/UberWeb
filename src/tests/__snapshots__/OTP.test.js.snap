// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sign up otp form render correctly 1`] = `
<div
  className="signupForm"
>
  <div
    className="col"
  >
    <p>
      Enter the 4-digit code sent to you at 
      123 456 7890
      . 
    </p>
    <p
      className="clickable-line"
      onClick={[MockFunction]}
    >
      Did you enter the correct number?
    </p>
    <div
      className="otp-row"
    >
      <input
        autoComplete="off"
        autoFocus={true}
        className="otp-box"
        key="0"
        maxLength="1"
        name="otp"
        onChange={[Function]}
        type="text"
        value=""
      />
      <input
        autoComplete="off"
        autoFocus={false}
        className="otp-box"
        key="1"
        maxLength="1"
        name="otp"
        onChange={[Function]}
        type="text"
        value=""
      />
      <input
        autoComplete="off"
        autoFocus={false}
        className="otp-box"
        key="2"
        maxLength="1"
        name="otp"
        onChange={[Function]}
        type="text"
        value=""
      />
      <input
        autoComplete="off"
        autoFocus={false}
        className="otp-box"
        key="3"
        maxLength="1"
        name="otp"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <p
      className="error-msg"
    />
    <p
      className="clickable-line"
    >
      I didn’t receive code.
    </p>
    <button
      className="round-btn"
      onClick={[Function]}
    >
      →
    </button>
  </div>
</div>
`;
